src/time/TimeGrids.cpp

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <Eigen/Core>
+#include <numeric>
+#include <vector>
+
+#include "TimeGrids.hpp"
+#include "utils/assertion.hpp"
+
+namespace precice::time {
+
+void TimeGrids::saveTimeGridOfCouplingData(const DataMap &cplData, std::vector<int> dataIDs, bool reducedTimeGrid)
+{
+  for (int dataID : dataIDs) {
+    Eigen::VectorXd timeGrid = cplData.at(dataID)->timeStepsStorage().getTimes();
+    if (reducedTimeGrid) {
+      _timeGrids.insert(std::pair<int, Eigen::VectorXd>(dataID, timeGrid.tail<1>()));
+    } else {
+      _timeGrids.insert(std::pair<int, Eigen::VectorXd>(dataID, timeGrid));
+    }
+  }
+}
+
+Eigen::VectorXd TimeGrids::getTimeGrid(int dataID) const
+{
+
+  if (_timeGrids.count(dataID)) {
+    PRECICE_ASSERT("there does not exists a stored time grid corresponding to this dataID");
+  }
+
+  return _timeGrids.at(dataID);
+}
+
+void TimeGrids::moveTimeGridToNewWindow(const DataMap &cplData)
+{
+  for (auto &pair : _timeGrids) {
+    if (pair.second.size() == 1) {
+      _timeGrids.at(pair.first) = cplData.at(pair.first)->timeStepsStorage().getTimes().tail<1>();
+    } else {
+      int dataID = pair.first;
+      // Only way to access the first time stamp is through the whole vector
+      Eigen::VectorXd newtimeGrid = cplData.at(dataID)->timeStepsStorage().getTimes();
+      double          newTimesMin = newtimeGrid(0);
+      double          newTimesMax = newtimeGrid(newtimeGrid.size() - 1);
+
+      Eigen::VectorXd timeGrid    = pair.second;
+      double          oldTimesMin = _timeGrids.at(dataID)(0);
+      double          oldTimesMax = _timeGrids.at(dataID)(timeGrid.size() - 1);
+
+      // transform the time to the new time grid
```

Comment:
I have added an explanation here of what this does. I decided to add the reason of why we need it to the moveTimeGridToNewWindow, since this is the main thing that this function does.